feat(api): Adds logic to filter nested results

ERosendo opened the pull request:
As suggested in [#4054 (comment)](https://github.com/freelawproject/courtlistener/issues/4054#issuecomment-2491862877), this PR introduces a new custom filter, `filter_nested_results`, to the PartyFilter and AttorneyFilter classes. This filter leverages `prefetch_related` to refine the filtering of `Role` records based on other API request filters. The filtered `Role` queryset is stored in the `filtered_roles` attribute.

The corresponding serializers have been tweaked to use the `filtered_roles` attribute when available, ensuring that API responses accurately reflect the filtered results.

Also, The API documentation has been updated to document the new filter and its impact on API requests.

![image](https://github.com/user-attachments/assets/62d88e0c-1926-4c93-928f-c277844d29bb)


![image](https://github.com/user-attachments/assets/59b1cff2-0e37-455d-ab1f-ec19039bfcd1)


fixes #4054

mlissner commented on cl/api/templates/pacer-api-docs-vlatest.html:
```diff
@@ -201,7 +201,8 @@ <h2 id="party-endpoint">Parties <small> — <code>{% url "party-list" version=ve
   <p>This API can be filtered by docket ID to show all the parties for a particular case.
   </p>
-  <p class="alert alert-warning"><i class="fa fa-warning"></i> <strong>Listen Up:</strong> Filters apply to this endpoint, not the data nested within it. Therefore, each party returned by this API will list all the attorneys that have represented them in any case, even if the parties themselves are filtered to a particular case.
+  <p class="alert alert-warning"><i class="fa fa-warning"></i> <strong>Listen Up:</strong> Filters apply to this endpoint only affect the top-level data, not the data nested records within it. Therefore, each party returned by this API will list all the attorneys that have represented them in any case, even if the parties themselves are filtered to a particular case. <br><br>
```
```suggestion
  <p class="alert alert-warning"><i class="fa fa-warning"></i> <strong>Listen Up:</strong> Filters applied to this endpoint only affect the top-level data, not the data nested records within it. Therefore, each party returned by this API will list all the attorneys that have represented them in any case, even if the parties themselves are filtered to a particular case. <br><br>
```

mlissner reviewed (commented):
One little comment on documentation. There's a bit more repetition in the approach than seems ideal, but I suppose removing the abstraction would probably make it even more abstract. I think if removing the repetition isn't possible, more comments might be useful. 

Also, it looks like each of the `filter_roles` methods has a list of expected keys that need to get cleaned up (in the `cleaned_key.replace` methods. Should we put a warning somewhere else in the code that if you add a filter to that place, you'll need to update this area of the code too? 

Tricky stuff.

ERosendo commented:
A SOLR test is failing. Maybe it senses the impending doom and is acting up

mlissner commented:
Just remove the test!

albertisfu commented on cl/api/tests.py:
```diff
@@ -1187,40 +1224,119 @@ async def test_attorney_filters(self) -> None:
+            2,
+            msg=f"Expected 2, but got {len(results[0]['parties_represented'])}.\n\nr.data was: {r.data}",
+        )
+
```
I observed a strange filtering behavior, which can be reproduced in this test by inserting the following lines here:

```suggestion
            await sync_to_async(RoleFactory.create)(
            party=party, docket=self.docket, attorney=self.attorney_2
        )
        self.q = {"docket__id": self.docket_2.pk}
        r = await self.async_client.get(self.path, self.q)
        results = r.data["results"]
        self.assertEqual(
            len(results),
            1,
            msg=f"Expected 1, but got {len(results)}.\n\nr.data was: {r.data}",
        )
```

It adds a new role involving `self.docket`, so when applying the filter `"docket__id": self.docket_2.pk` the new role doesn't modify the results from this filter and continues returning only one result.

However, the filter below returns different results:

```
 {
            "parties_represented__docket__id": self.docket_2.pk,
            "filter_nested_results": True,
        }
```

In this case, it matches the attorney linked to `self.docket_2.pk` and also matches `self.attorney_2`. However, this is odd because the `Role` created is linked to `self.docket`. This can be confirmed by checking `self.attorney_2` `parties_represented` key, which is empty. So, while the nested filter is working, the parent filter does not appear to be applied correctly. Is this behavior expected?

ERosendo replied:
@albertisfu I apologize for the misunderstanding. It's my fault. When you asked about the `docket` lookup in the `PartyFilter` class, I was under the impression that it was using the `role` table. However, it actually uses the `party_types` table for filtering.

the `parties_represented__docket__id` lookup should return attorneys whose represented parties are linked to docket_2 through the `partyType` model. The addition of the new role is causing the test to fail because the party record is linked to `docket_2`(check line 1277).

ERosendo replied:
> this is the expected behavior because, even though Attorney 30 doesn’t have a Role directly related to docket_id=8, it's shown because party=15 is related to docket_id=8 in a different Role. Due to the filter uses the PartyType table for filtering, correct?

This isn't quite right. You're getting those attorneys because Parties `15` and `14` are linked to Docket `8` through the `partyType` table. Once you use `parties_represented__docket__id`, you're no longer looking at the roles table. 

Can you please verify that Parties 14 and 15 are linked to Docket 8?

albertisfu replied:
Got it! Yeah Parties 14 and 15 are related to Docket 8 through `partyType` instances. 

Is that also why `filter_nested_results=true` is unable to filter `parties_represented` in attorney 30, because party 15 is linked to docket_id 8,  correct?

ERosendo replied:
> Is that also why filter_nested_results=true is unable to filter parties_represented in attorney 30, because party 15 is linked to docket_id 8, correct?

That's correct!

albertisfu commented on cl/api/utils.py:
```diff
@@ -91,6 +92,96 @@ def to_html(self, request, queryset, view):
+                continue
+
+            cleaned_key = filter_key
+            # Add base model prefix for fields in Meta class
+            for basic_field, _ in self._meta.fields.items():
+                if cleaned_key.startswith(basic_field):
+                    cleaned_key = (
+                        f"{base_model_prefix}{LOOKUP_SEP}{filter_key}"
+                    )
```
I noticed that this block of code is used to convert the base filters in the model (e.g., `id`, `date_created`, `date_modified`, `name` in the `attorney` model) into fields supported for nested lookups, such as` attorney__id`, `attorney__date_created`, `attorney__date_modified`, and `attorney__name`. I understand that this is necessary to prevent errors related to base fields not being supported for nested lookups.

However, if we apply the validation suggested in `filter_roles()` (in the other comment) to pass only valid `Role` fields to the query, could we remove these lines?

Since the base filters applied to nested instances might be redundant. For example, the attorney endpoint returns attorneys with nested `parties_represented`, which are `Roles` related to the same attorney. So any valid `attorney__` filter would already match all nested `Roles` if it matched the parent attorney?  And I think the same apply in the party endpoint.

ERosendo replied:
> the base filters applied to nested instances might be redundant

You're right. I'll add logic to ignore filter from the base model.

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -249,18 +251,32 @@ class Meta:
+    # RelatedFilters or custom labels in this class to avoid unexpected
+    # behavior.
+    join_table_cleanup_mapping = {
+        "attorney__docket": "docket",
```
Would it make sense to also add support for the `role` field here?
`"attorney__role": "role"`

ERosendo replied:
The specified lookup is not valid. The `PartyFilter` class relies on a related filter for attorney lookups, limiting valid attributes to those defined in the AttorneyFilter class. In our specific case, these attributes are restricted to `id`, `date_created`, `date_modified`, `name`, `docket`, and `parties_represented`.

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -249,18 +251,32 @@ class Meta:
+    # behavior.
+    join_table_cleanup_mapping = {
+        "attorney__docket": "docket",
+        "attorney__parties_represented": "party",
```
Since the filter `"attorney__parties_represented"` is available, would it make sense to add the `party` field to `AttorneyRoleSerializer` to make it easier to understand why those results matched the filter?

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -270,8 +286,21 @@ class Meta:
+        if not value:
+            return qs
+
+        role_filters = self.get_filters_for_join_table(name)
```
I found that when passing an invalid filter like:

`?test__contains=Richard&filter_nested_results=true`
or
`?parties_represented__attorney_name=lorem&filter_nested_results=true`

Returns an error like:
Cannot resolve keyword 'attorney_name' into field. Choices are: attorney, attorney_id, date_action, docket, docket_id, id, party, party_id, role, role_raw

Or
`?attorney__test=23&filter_nested_results=true`

Unsupported lookup 'test' for ForeignKey or join on the field not permitted.

This is because the invalid filter is being passed to:
`queryset=Role.objects.filter(**role_filters)`

I believe this can be solved by verifying if the filters returned by `get_filters_for_join_table` are valid `Role` fields.

We could do something like: 
`valid_role_fields = set(field.name for field in Role._meta.get_fields())`

And use the returned fields to filter out filters returned by `get_filters_for_join_table`

So invalid filters passed could be just ignored?

ERosendo replied:
> So invalid filters passed could be just ignored?

Thanks for pointing that out, @albertisfu!

There's a more elegant way to handle invalid filters in [Django-filter](https://github.com/carltongibson/django-filter/tree/main). The [BaseFilter](https://github.com/carltongibson/django-filter/blob/2494df96c6387a9fa411fcb00b696b15dfd9216b/django_filters/filterset.py#L186) class utilizes a form for validation. We can leverage this to filter out any invalid filters before applying them to the queryset.

Here's a link to the relevant section of the `filter_queryset` method in django-filter:

https://github.com/carltongibson/django-filter/blob/2494df96c6387a9fa411fcb00b696b15dfd9216b/django_filters/filterset.py#L228-L237

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -284,6 +313,19 @@ class AttorneyFilter(NoEmptyFilterSet):
+    # behavior.
+    join_table_cleanup_mapping = {
+        "parties_represented__docket": "docket",
+        "parties_represented__attorney": "attorney",
```
Should we include a mapping to `role` to?

`"parties_represented__role": "role"`

So role also available for filtering the nested results?

ERosendo replied:
[this comment](https://github.com/freelawproject/courtlistener/pull/4729/files#r1862587924) is also applicable here. To enable `role` lookup, we would need to update the class used in the `RelatedFilter` definition (specifically, the `PartyFilter` class in this instance)

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -284,6 +313,19 @@ class AttorneyFilter(NoEmptyFilterSet):
+    join_table_cleanup_mapping = {
+        "parties_represented__docket": "docket",
+        "parties_represented__attorney": "attorney",
+        "parties_represented": "parties",
```
I think the value in this line should be `party` ?

```suggestion
        "parties_represented": "party",
```
Otherwise I noticed that a filter like: `?parties_represented=8&filter_nested_results=true`

Raise: 
`Cannot resolve keyword 'parties' into field. Choices are: attorney, attorney_id, date_action, docket, docket_id, id, party, party_id, role, role_raw`

ERosendo replied:
Thanks for catching this.

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -293,3 +335,16 @@ class Meta:
             "date_modified": DATETIME_LOOKUPS,
             "name": ALL_TEXT_LOOKUPS,
         }
+
+    def filter_roles(self, qs, name, value):
+        if not value:
+            return qs
+
+        role_filters = self.get_filters_for_join_table(name)
+
+        prefetch = Prefetch(
+            name,
+            queryset=Role.objects.filter(**role_filters),
+            to_attr=f"filtered_{name}",
+        )
+        return qs.prefetch_related(prefetch)
```
I noticed that this block of code is the same as in `PartyFilter`. Would it be possible to move it to a common place? Maybe within `FilterManyToManyMixin` or as a shared helper method?

ERosendo replied:
> Maybe within FilterManyToManyMixin or as a shared helper method?

While refactoring for common logic is generally a good idea, I believe a separate mixin for role filtering would provide a cleaner solution. This would keep the `FilterManyToManyMixin` focused on its core functionality and avoid potential conflicts or overcomplication.

ERosendo commented:
@albertisfu Thanks for your feedback on the initial approach. I've made the following adjustments based on your suggestions:

1. **Filtering and Validation:**
   
   In your feedback, you highlighted two issues with my original approach:

   - The mixin wasn't validating the filters in the query string.
   - The final set of filters included unnecessary lookups because the mixin was pulling in fields from the base model. You suggested we should only focus on the fields related to the nested data.

    While reviewing the Django-filters code, I noticed that the base FilterSet class uses a Django form to validate query string data. I thought we could leverage the cleaned_data property of this form to address the first issue.

    To address the second issue, I initially planned to use the `related_filters` property. This property specifically holds filters for nested lookups. However, I realized that the `cleaned_data` property wasn't sufficient for deeply nested lookups like `parties_represented__docket__docket_entries__recap_documents__id`. While it effectively ignored invalid filters, it didn't provide a straightforward way to extract these nested lookups.

     I revisited my [comment](https://github.com/freelawproject/courtlistener/issues/4054#issuecomment-2491862877) on the issue and checked the Django-filters repository discussions. I found that each FilterSet has a useful `qs` property. This property stores a filtered queryset based on the form's cleaned data. The mixin now uses the qs property to find the necessary primary keys of the records that should be included in the API response. This happens when the filter_nested_results parameter is used, and the filterset's lookups are simplified to only __in lookups.


2. **Addressing Filtering Inconsistencies**

    You brought up an interesting issue with the filtering behavior. [I explained why the behavior was correct](https://github.com/freelawproject/courtlistener/pull/4729#discussion_r1870463151), but in doing so, I realized that the party endpoint is showing two join tables from different many-to-many relationships. Specifically, it's displaying the role table in the `attorneys` field and party types in the `party_types` field.

    To address this, I updated the custom method to apply filtering to both nested relationships. Previously, the filter was only applied to the role table. I've also updated the tests to verify that the new behavior correctly filters both nested objects within the party endpoint.

albertisfu commented on cl/api/utils.py:
```diff
@@ -91,6 +93,135 @@ def to_html(self, request, queryset, view):
+        return join_table_key
+
+    def get_filters_for_join_table(
+        self: FilterSet, name: str
```
It seems the argument name is not being used. Should we remove it?

ERosendo replied:
> It seems the argument name is not being used

You're right!. We can remove it

albertisfu commented on cl/people_db/filters.py:
```diff
@@ -293,3 +365,31 @@ class Meta:
+        prefetch = Prefetch(
+            name,
+            queryset=Role.objects.filter(**role_filters),
+            to_attr=f"filtered_{name}",
```
Could there be a case where `to_attr` is different from `filtered_roles`? If so, perhaps make it fixed without using `{name}`. Otherwise, adjust accordingly in `get_parties_represented`.

ERosendo replied:
> Could there be a case where to_attr is different from filtered_roles?

@albertisfu We can use a different name. I simply used `filtered_roles` because we're using the `Role` model to construct the prefetch query. Would `filtered_parties_represented` be a more suitable alternative?

albertisfu replied:
Yeah, `filtered_roles` is fine, or `filtered_parties_represented`. I just wondered if it could change dynamically according to `name`, since it's set as `filtered_{name}`. However, it seems it will always be `filtered_roles`, so maybe there's no need to use `name` in it?

ERosendo replied:
Got it! I'll use a static name. Thanks!

albertisfu replied:
Great! :shipit:

albertisfu reviewed (commented):
Thanks @ERosendo changes look great! Just a couple of minor comments.

albertisfu reviewed (approved):
Thanks @ERosendo! This looks great to me now :shipit:

ERosendo commented:
after merging this PR let's also consider merging https://github.com/freelawproject/courtlistener/pull/4745

mlissner commented:
Thank you both. This one was tricky!
